### ISSUE TITLE
MySQL Connection Pool size and close/timeout issue

### ISSUE THREAD

**Issue by sagarch88** [typeorm/typeorm#2441]
**Issue type:**

[x ] question
[ ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ x] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[ ] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[x ] `latest`
[ ] `@next`
[ ] `0.x.x` (or put your version here)

1. Does TypeORM for MySQL by default uses connection pool? (I believe it does)
2. How do I specify the connectionPool limit in TypeORM for MySQL?
3. Should I manually release the connection from connectionPool? If yes, how do I release the connection?

I have followed the documentation for the implementation of TypeORM. Following is the ormconfig.json
`[

    {
        "name": "default",
        "type": "mysql",
        "host": process.env.DATABASE_HOST,
        "port": process.env.DATABASE_PORT,
        "username": process.env.DATABASE_USER,
        "password": process.env.DATABASE_PASSWORD,
        "connectTimeout": 1500,
        "database": process.env.DATABASE_NAME,
        "entities": [
            "src/models/entities/**/*.ts"
        ],
        "logging": true
    }
]`

I am having issues with TypeORM with MySQL.

1. Immediately after NodeJS application is started , the MySQL query takes 1 second to respond. If no more requests are received to the server, the connection to MySQL is still active, but under sleep state(By running `show processlist;` on MySQL - application user is still active.). After the idle time of around 10 minutes, If there is other request, I get the response in around 20 seconds for the 1st time and then immediately, for subsequent requests.

I have traced the issue to the following line of code in **my Application**:
        `await getConnectionManager().get().transaction        // Takes around 20sec.`
I have tried with `getManager().transaction` too but the issue persists. By splitting the above statement into 2 lines, 
    `const connectionManager = getManager();`
    `connectionManager.transaction();`, getting the transaction is taking the time.

2. The connections are never closed on the MySQL server side. In the following image, where application user is `unified`. 20+ users are under sleep connection even when there were no requests to the server.
![screen shot 2018-07-02 at 9 56 00 am](https://user-images.githubusercontent.com/23547857/42145906-c2c052cc-7de1-11e8-9e1f-b084790364d2.png)

There are sleep state, even when the application is not receiving any requests. Shouldn't the connection pool maintain just one connection and close others if not needed. - (_Reference to 3 question above_)

@pleerock - This is my first post on GitHub forum. I hope I have explained the questions/issues in detail. Please let me know if any other details are required.

**Comment by pleerock** [comment]
> Does TypeORM for MySQL by default uses connection pool? (I believe it does)

yes

> How do I specify the connectionPool limit in TypeORM for MySQL?

using `extra` options and sent options from underlying driver

> Should I manually release the connection from connectionPool? If yes, how do I release the connection?

no, you shouldn't 

TypeORM creates a connection pool and uses connection for a single operation (find, save, remove, etc.) - it takes the connection, uses it and releases it. That's it. If all connections are busy it waits until any of them is released and then use it. You should not have issues you are talking about. I guess its something in your setup. Please provide some minimal reproduction code that shows the problem you have.

**Comment by sagarch88** [comment]
1. For using `extra` option can you please provide link to its corresponding documentation?

2. My code is divided into 2 layers - `serviceLayer` & `repositoryLayer`

Following code shows my minimal code and flow of logic.

On `service` layer:

    import { injectable, inject } from "inversify";
    import { EntityManager } from "typeorm";
    import TYPES from "../types";

    import { ServiceBase } from "./servicebase";
    import { LeadDTO } from "../models/entities/Lead";
    import { LeadRepository } from "../repositories/lead.repository";

    export interface LeadService extends ServiceBase {
        getLead(code: string): Promise<LeadDTO>;
    }

    @injectable()
    export class LeadServiceImpl implements LeadService {
        @inject(TYPES.LeadRepository)
        private leadRepositoryDb: LeadRepository;

        public async getLead(code: string): Promise<LeadDTO> {
            try {
                 const lead: LeadDTO = await getConnectionManager().get().transaction(async transManager => {
                this.leadRepositoryDb.setEntityManager(this.transManager);
                const leadDTO = await this.leadRepositoryDb.find(code).then((a) => {
                    // Convert to DTO and return
                    return a;
                });
                return leadDTO;
            });
            } catch (error) {
                throw error;
            }
        }
    }


Where I am passing **same** `entityManager` received from 'Service' layer to `repo` layer - (So that we can rollback the whole transaction when working on multiple `repositories` files from same `Service` layer)

On `repository` layer:

    import { injectable } from "inversify";
    import { getRepository, EntityManager } from "typeorm";

    import { RepositoryBase } from "./repositorybase";
    import { LeadDTO, Lead } from "../models/entities/Lead";

    export interface LeadRepository extends RepositoryBase<LeadDTO> {
        find(code: string): Promise<Lead>;
    }

    @injectable()
    export class LeadRepositoryImpl implements LeadRepository {
        private entityManager: EntityManager;

        public setEntityManager(entityManager: EntityManager): void {
            try {
                this.entityManager = entityManager;
            }
            catch (error) {
                throw error;
            }
        }

        public getEntityManager(): EntityManager {
            try {
                return this.entityManager;
            }
            catch (error) {
                throw error;
            }
        }

        public async find(code: string): Promise<Lead> {
            try {
                return await getRepository(Lead).findOne(code);
            }
            catch (error) {
                throw error;
            }
        }

        public async create(lead: Lead): Promise<Lead> {
            try {
                return await this.entityManager.save(lead);
            } catch (error) {
                throw error;
            }
        }
    }

I am using the `entityManager` to `save/update`. Following has both `GET` via `getRepository(Lead)` and `save/update` via `entityManager`.

And finally `lead.entity` file

    import { Entity, PrimaryColumn, Column, UpdateDateColumn, CreateDateColumn, BeforeInsert } from "typeorm";
    import { Length, IsEmail, IsNumberString } from "class-validator";

    export interface LeadDTO {
        code: string;
        name: string;
    }

    @Entity("lead")
    export class Lead implements LeadDTO {
        @PrimaryColumn({
            type: "varchar",
            width: 50,
            unique: true
        })
        public code: string;

        @Length(50, 500)
        @Column({
            type: "varchar",
            width: 500
        })
        public name: string;
    }

In the above code, I don't `release` or `close` the connection manually assuming `TypeORM` should close it. But I am still having `active` connections under `sleep state` on MySQL server and delay in initial response for some weird reason.

@pleerock, I have followed the above code pattern through out the application. Please check and let me know if any more details are required.

**Comment by pleerock** [comment]
https://github.com/mysqljs/mysql#pool-options

in typeorm you pass those options in `extra` of connection options:

```ts
{
  ...,
  "extra": { connectionLimit: 10, ... }
}
```

**Comment by pleerock** [comment]
Can you please fork https://github.com/typeorm/typescript-example change it to mysql and check if you have same issue? If issue is in typeorm we need to make a minimal possible reproduction

**Comment by sagarch88** [comment]
I have a weird observation.

I added the `connectionLimit` and deployed the Application. And the above mentioned issue does not occur anymore. I believe, if `connectionLimit` is not mentioned, the `default` 10 connectionLimit is getting ignored. (Not sure though). Please check and if incorrect, please close the issue.

Thank You.

**Comment by michaelchiche** [comment]
I am having the same issue, by running this:

```ts
getConnection(process.env.NODE_ENV).getRepository(Discussion)
```
where Discussion is an entity.

**Comment by sagarch88** [comment]
After the deployment on 2 different MySQL servers, I have another observation.

1st MySQL server (Dev): It is self hosted on EC2 where I am facing connection delay for initial first request.

2nd MySQL server (Prod): It is hosted on AWS RDS and the same code base is not having the issues.

I am not sure what configs are required on MySQL DB server.

@pleerock @michaelchiche , hope above info gives some clue to this weird issue.

**Comment by michaelchiche** [comment]
For me, it the regular config, running inside a docker container

**Comment by michaelchiche** [comment]
FWIW, I am pretty sur it comes from how TypeORM manages the connections. Because I have a grafana plugged to my database with prometheus, to gather metrics for monitoring every 5 seconds. And there is no problem on that side.

**Comment by michaelchiche** [comment]
I think I found where the issue is: ts-node. Are you using ts-node alors on your dev server? @sagarch88

**Comment by sagarch88** [comment]
Sorry. Didn't get what you mean by `alors`? 

My node application is running on docker container, in the compiled JS code and not on ts-node. With application running on ts-node locally also, I have the same above issue with self-hosted MySQL DB on EC2.

**Comment by michaelchiche** [comment]
Damn auto correct. I meant also

**Comment by sagarch88** [comment]
Yes. The issue exists both with `ts-node` and with the `complied JS` code.

**Comment by Slessi** [comment]
I too have this problem. I increased the connectionLimit to confirm, and I now see the list growing endlessly.

Similar to @sagarch88 I use `ts-node` in development but `node` in production, and the problem occurs irrespective of which I use.

![screen shot 2018-08-22 at 14 55 39](https://user-images.githubusercontent.com/3223500/44464703-c5f07c80-a61b-11e8-92e8-348071077478.jpg)

**Comment by Slessi** [comment]
I found my problem. Not sure if it is also what caused the OP's problem.

> tl;dr
> when using `const qRunner = connection.createQueryRunner()` you *must* at some point call `qRunner.release()` yourself to avoid running out of free connections

I followed the documentation for transactions here: [docs/transactions.md](https://github.com/typeorm/typeorm/blob/master/docs/transactions.md#using-queryrunner-to-create-and-control-state-of-single-database-connection)

Specifically, I was opting for _"Using QueryRunner to create and control state of single database connection"_.

If you look at [src/driver/mysql/MysqlQueryRunner.ts](https://github.com/typeorm/typeorm/blob/master/src/driver/mysql/MysqlQueryRunner.ts) you will see that neither `commitTransaction` nor `rollbackTransaction` releases the connection for you, so you will need to release it yourself, otherwise you will eventually fill up your pool (default limit 10) with unreleased connections.

The documentation I linked above does not mention anything about needing to manually release the connection when using query runner, and the [docs/query-runner.md](https://github.com/typeorm/typeorm/blob/master/docs/query-runner.md) doc is marked as `TBD`.

@pleerock Maybe you can shed some light?
Is there a bug, incomplete documentation or am I just stupid?

**Comment by sagarch88** [comment]
@Slessi,

As mentioned above, for me the `running out of connections issue` was resolved by manually adding the `connectionLimit` under `extra` in `ormconfig.json`. Can you please add and check if the issue is resolved?

**Comment by Slessi** [comment]
No that was not my issue.

When I don't specify I get a default of 10 which works for me and editing to any other value works as well.

My issue was that the created connections for the pool were not being released.

**Comment by michaelchiche** [comment]
I am still struggling with this issue, even after adding:
```js
extra: {
  connectionLimit: 50,
}
```

If no request are made for about ten minutes, I have to wait about 2minutes before the connection is "live" again

**Comment by ajkamel** [comment]
I too had this issue but in when following the original documentation for using `queryRunner.startTransaction();`  The problem was that the documentation example never called `queryRunner.release();` Which led to us exhausting the number of connections available.

**Comment by pleerock** [comment]
Documentation was updated @Slessi by #2681 . @sagarch88 did you resolve your issue by manually closing opened connections?

**Comment by Kononnable** [comment]
Closing as no response from the reporter.

**Comment by ioclaudio** [comment]
> TypeORM creates a connection pool and uses connection for a single operation (find, save, remove, etc.) - it takes the connection, uses it and releases it. That's it. If all connections are busy it waits until any of them is released and then use it. You should not have issues you are talking about. I guess its something in your setup. Please provide some minimal reproduction code that shows the problem you have.

Hi,
pleerock we are experiencing the same problem.
We don't have specified the option: extra: {  connectionLimit: 50, }
So I think that the default value is used.

The service uses code similar to this to manage transactions:
```
const connection = getConnection();
const queryRunner = connection.createQueryRunner();
await queryRunner.connect();
await queryRunner.startTransaction();
 
try {
	//... many queries
	await queryRunner.commitTransaction();

} catch (err) {
	// Since we have errors lets rollback changes we made
	console.info(err);
	await queryRunner.rollbackTransaction();
	res.status(409).send();
	return;
} finally {
	// Release query runner
	await queryRunner.release();
}
```

However, the number of connections increases over time.
Why are the connections never closed and why they never expire?

cld

**Comment by a7urag** [comment]
@ioclaudio try to return(from catch) after releasing the connection(in finally)

**Comment by 0xckylee** [comment]
@ioclaudio were you able to resolve this issue by adding `return` after releasing the connection?
cc. @a7urag
